src/bp/core/services/bot-service.ts

```diff
@@ -451,4 +465,75 @@ export class BotService {
     BotService._mountedBots.forEach((isMounted, bot) => isMounted && bots.push(bot))
     return bots
   }
+
+  public async listRevisions(botId: string): Promise<string[]> {
+    const globalGhost = this.ghostService.global()
+
+    let stageID = ''
+    if (await this.workspaceService.hasPipeline()) {
+      const botConfig = await this.configProvider.getBotConfig(botId)
+      stageID = botConfig.pipeline_status.current_stage.id
+    }
+
+    const revisions = await globalGhost.directoryListing(REVISIONS_DIR)
+
+    return revisions.filter(rev => rev.includes(botId) && rev.includes(stageID)).sort((revA, revB) => {
```

Comment:
That's potentially unsecure in the following scenarios:

bot called: `hell` can list/restore revisions of all bots that contain `hell` in it, such as `ashell` and `hello`
same for stage names (although less likely)

I'd do a full name check, i.e. startsWith(`${botId}::${stageID}`)